File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/spots/entryspot/BingoEntrySpot.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.entryspot;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkActorCreateResponse;
import systems.zlink.framework.spots.ZLinkEntrySpot;
import systems.zlink.framework.spots.ZLinkEntrySpotContext;
import systems.zlink.framework.spots.ZLinkSpotManager;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.server.configuration.SampleTimings;
import systems.zlink.samples.bingo.server.play.domain.bingo.BingoRoomModels;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.actors.PlayerActor;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class BingoEntrySpot implements ZLinkEntrySpot<PlayerActor> {
    private static final Logger logger = LoggerFactory.getLogger(BingoEntrySpot.class);

    private final ZLinkEntrySpotContext context;
    private final ZLinkSpotManager spots;

    public BingoEntrySpot(ZLinkEntrySpotContext context, ZLinkSpotManager spots) {
        this.context = context;
        this.spots = spots;
    }

    @Override
    public ZLinkEntrySpotContext context() {
        return context;
    }

    @Override
    public CompletionStage<ZLinkActorCreateResponse> onCreateActor(
            PlayerActor actor, ZLinkMessage createRequest) {
        Messages.EnsurePlayerActorReq request =
                createRequest.decode(Messages.EnsurePlayerActorReq.class);
        actor.setDisplayName(request.getDisplayName());
        return CompletableFuture.completedFuture(ZLinkActorCreateResponse.accept());
    }

    // --8<-- [start:doc-bingo-entry-destroy]
    @Override
    public CompletionStage<Void> onJoinedActor(PlayerActor actor) {
        if (actor.destroyAfterEntrySpotJoin()) {
            return context.destroyActor(actor)
                    .thenRun(
                            () ->
                                    logger.info(
                                            "bingo-lifecycle entry-destroy-complete actor={}",
                                            actor.actorId()));
        }
        return CompletableFuture.completedFuture(null);
    }

    // --8<-- [end:doc-bingo-entry-destroy]

    @Override
    public CompletionStage<Void> onLeaveActor(PlayerActor actor) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnectActor(PlayerActor actor) {
        actor.markDisconnected();
        return CompletableFuture.completedFuture(null);
    }

    public CompletionStage<Messages.ObserveBingoEventsRes> observeEvents(
            PlayerActor actor, Messages.ObserveBingoEventsReq request) {
        String observerSpotId = "observe:" + request.getRoomId() + ":" + actor.actorId();
        BingoRoomModels.BingoRoomSettings settings =
                BingoRoomModels.BingoRoomSettings.createObserver(
                        request.getRoomId(), actor.actorId(), SampleTimings.DrawPeriod.toMillis());
        Messages.BingoRoomSettingsPayload settingsPayload =
                Messages.BingoRoomSettingsPayload.newBuilder()
                        .setRoomName(settings.roomName())
                        .setMode(settings.mode())
                        .setRequiredPlayers(settings.requiredPlayers())
                        .setMaxDrawNumber(settings.maxDrawNumber())
                        .setPurpose(settings.purpose())
                        .setObservedRoomId(settings.observedRoomId())
                        .build();
        return spots.getOrCreate(observerSpotId, SampleNames.RoomSpotType)
                .inMesh(SampleNames.Mesh)
                .request(BingoMessages.bingoRoomCreateReq(settingsPayload))
                .submit()
                .thenApply(
                        ignored -> {
                            actor.context()
                                    .joinSpot(
                                            observerSpotId,
                                            BingoMessages.bingoRoomJoinReq(
                                                    request.getRoomId(),
                                                    actor.actorId(),
                                                    actor.displayName(),
                                                    true))
                                    .defer();
                            return BingoMessages.observeBingoEventsRes(true);
                        });
    }
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/spots/entryspot/BingoEntrySpot.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.entryspot

import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpot
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkActorCreateResponse
import systems.zlink.framework.spots.ZLinkEntrySpotContext
import systems.zlink.framework.spots.ZLinkSpotManager
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.server.play.domain.bingo.BingoRoomSettings
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors.PlayerActor
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomCreateReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomJoinReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomSettingsPayload
import systems.zlink.samples.kotlin.bingo.shared.contracts.EnsurePlayerActorReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.ObserveBingoEventsReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.ObserveBingoEventsRes

class BingoEntrySpot(
    override val context: ZLinkEntrySpotContext,
    private val spots: ZLinkSpotManager,
) : ZLinkSuspendingEntrySpot<PlayerActor>() {
    private val logger = LoggerFactory.getLogger(BingoEntrySpot::class.java)

    override suspend fun onCreateActorSuspending(
        actor: PlayerActor,
        createRequest: ZLinkMessage,
    ): ZLinkActorCreateResponse {
        val request = createRequest.decode(EnsurePlayerActorReq::class.java)
        actor.setDisplayName(request.displayName)
        return ZLinkActorCreateResponse.accept()
    }

    // --8<-- [start:doc-bingo-entry-destroy]
    override suspend fun onJoinedActorSuspending(actor: PlayerActor) {
        if (actor.destroyAfterEntrySpotJoin) {
            context.destroyActor(actor).await()
            logger.info("bingo-lifecycle entry-destroy-complete actor={}", actor.actorId())
        }
    }

    // --8<-- [end:doc-bingo-entry-destroy]

    override suspend fun onLeaveActorSuspending(actor: PlayerActor) {}

    override suspend fun onDisconnectActorSuspending(actor: PlayerActor) {
        actor.markDisconnected()
    }

    suspend fun observeEvents(
        actor: PlayerActor,
        request: ObserveBingoEventsReq,
    ): ObserveBingoEventsRes {
        val observerSpotId = "observe:${request.roomId}:${actor.actorId()}"
        val settings =
            BingoRoomSettings.createObserver(
                request.roomId,
                actor.actorId(),
                SampleTimings.DrawPeriod.toMillis(),
            )
        val settingsPayload =
            BingoRoomSettingsPayload(
                mode = settings.mode,
                roomName = settings.roomName,
                requiredPlayers = settings.requiredPlayers,
                maxDrawNumber = settings.maxDrawNumber,
                purpose = settings.purpose,
                observedRoomId = settings.observedRoomId ?: "",
            )
        spots
            .getOrCreate(observerSpotId, SampleNames.RoomSpotType)
            .inMesh(SampleNames.Mesh)
            .request(BingoRoomCreateReq(settingsPayload))
            .submit()
            .await()
        actor
            .context()
            .joinSpot(
                observerSpotId,
                BingoRoomJoinReq(request.roomId, actor.actorId(), actor.displayName, true),
            )
            .defer()
        return ObserveBingoEventsRes(true)
    }
}
